File: host/src/main/java/com/reservation/host/accommodation/controller/AccommodationController.java
package com.reservation.host.accommodation.controller;

import static com.reservation.common.response.ApiResponse.*;
import static com.reservation.host.auth.controller.AuthController.*;

import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import com.reservation.common.response.ApiResponse;
import com.reservation.commonauth.auth.annotation.LoginUserId;
import com.reservation.commonmodel.accommodation.AccommodationDto;
import com.reservation.host.accommodation.controller.dto.request.CreateAccommodationRequest;
import com.reservation.host.accommodation.controller.dto.request.UpdateAccommodationRequest;
import com.reservation.host.accommodation.service.AccommodationService;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/accommodation")
@Tag(name = "숙소 관리 API", description = "숙박 업체 숙소 관리 API입니다.")
@PreAuthorize(PRE_AUTH_ROLE_HOST)//✅숙박 업체만 접근 가능
@RequiredArgsConstructor
public class AccommodationController {
	private final AccommodationService accommodationService;

	@PostMapping
	@Operation(summary = "숙소 등록", description = "숙박 업체가 숙소를 등록합니다.")
	@ResponseStatus(HttpStatus.CREATED)
	public ApiResponse<Long> createAccommodation(@Valid @RequestBody CreateAccommodationRequest request,
		@LoginUserId Long hostId) {
		Long accommodationId = accommodationService.createAccommodation(request, hostId);
		return ok(accommodationId);
	}

	@PutMapping
	@Operation(summary = "숙소 수정", description = "숙박 업체가 숙소를 수정합니다.")
	public ApiResponse<Long> updateAccommodation(@Valid @RequestBody UpdateAccommodationRequest request,
		@LoginUserId Long hostId) {
		Long accommodationId = accommodationService.updateAccommodation(request, hostId);
		return ok(accommodationId);
	}

	@GetMapping
	@Operation(summary = "숙소 조회", description = "숙박 업체가 숙소를 조회합니다.")
	public ApiResponse<AccommodationDto> findHostAccommodation(@LoginUserId Long hostId) {
		AccommodationDto accommodation = accommodationService.findHostAccommodation(hostId);
		return ok(accommodation);
	}
}

File: accommodation-service/src/main/kotlin/com/msa/accommodationservice/accommodation/controller/AccommodationController.kt
package com.msa.accommodationservice.accommodation.controller

import com.msa.accommodationservice.accommodation.controller.request.AccommodationRegistrationRequest
import com.msa.accommodationservice.accommodation.controller.request.AccommodationUpdateRequest
import com.msa.accommodationservice.accommodation.controller.response.AccommodationInfoResponse
import com.msa.accommodationservice.accommodation.controller.response.AccommodationInfosResponse
import com.msa.accommodationservice.accommodation.service.AccommodationService
import com.msa.supportmodule.annotation.HostId
import com.msa.supportmodule.response.ApiResponse
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*


@RestController
@RequestMapping("/accommodations")
class AccommodationController(
    private val service: AccommodationService
) {

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun register(
        @HostId hostId: Long,
        @Valid @RequestBody request: AccommodationRegistrationRequest,
    ): ApiResponse<AccommodationInfoResponse> {
        val dto = request.toAccommodationDto(hostId)

        val accommodation = service.register(dto)

        val response = AccommodationInfoResponse(
            accommodationId = accommodation.id,
            name = accommodation.name,
            description = accommodation.description,
            contactNumber = accommodation.contactNumber,
            city = accommodation.city,
            address = accommodation.address,
            hostId = accommodation.hostId,
        )

        return ApiResponse.create(
            message = "숙소 정보를 등록 했습니다.",
            data = response
        )
    }

    @PatchMapping()
    fun update(
        @HostId hostId: Long,
        @Valid @RequestBody request: AccommodationUpdateRequest,
    ): ApiResponse<AccommodationInfoResponse> {
        val dto = request.toAccommodationDto(hostId)

        val accommodation = service.update(dto)
        val response = AccommodationInfoResponse(
            accommodationId = accommodation.id,
            name = accommodation.name,
            description = accommodation.description,
            contactNumber = accommodation.contactNumber,
            city = accommodation.city,
            address = accommodation.address,
            hostId = accommodation.hostId,
        )

        return ApiResponse.update(
            message = "숙소 정보를 수정 했습니다.",
            data = response
        )
    }

    // TODO: 서치 및 페이징 처리
    @GetMapping()
    fun getAccommodations(): ApiResponse<AccommodationInfosResponse> {
        val accommodations = service.getAccommodations()

        val responseInfos = accommodations.map { accommodation ->
            AccommodationInfoResponse(
                accommodationId = accommodation.id,
                name = accommodation.name,
                description = accommodation.description,
                contactNumber = accommodation.contactNumber,
                city = accommodation.city,
                address = accommodation.address,
                hostId = accommodation.hostId,
            )
        }
        val response = AccommodationInfosResponse(
            accommodations = responseInfos
        )

        return ApiResponse.read(
            message = "숙소 리스트 조회에 성공 했습니다.",
            data = response
        )
    }

    @GetMapping("{accommodationId}")
    fun getAccommodation(
        @PathVariable accommodationId: Long
    ): ApiResponse<AccommodationInfoResponse> {
        val accommodation = service.getAccommodation(accommodationId)
        val response = AccommodationInfoResponse(
            accommodationId = accommodation.id,
            name = accommodation.name,
            description = accommodation.description,
            contactNumber = accommodation.contactNumber,
            city = accommodation.city,
            address = accommodation.address,
            hostId = accommodation.hostId,
        )

        return ApiResponse.read(
            message = "숙소 정보를 조회에 성공했습니다.",
            data = response
        )
    }

    @GetMapping("me")
    fun getMyAccommodation(
        @HostId hostId: Long
    ): ApiResponse<AccommodationInfoResponse> {
        val accommodation = service.getAccommodationByHostId(hostId)
        val response = AccommodationInfoResponse(
            accommodationId = accommodation.id,
            name = accommodation.name,
            description = accommodation.description,
            contactNumber = accommodation.contactNumber,
            city = accommodation.city,
            address = accommodation.address,
            hostId = accommodation.hostId,
        )

        return ApiResponse.read(
            message = "숙소 정보를 조회에 성공했습니다.",
            data = response
        )
    }

}
